### ISSUE TITLE
[BUG] test_hash_groupby_approx_percentile_long_repeated_keys failed intermittently

### ISSUE THREAD

**Issue by pxLi** [NVIDIA/spark-rapids#4060]
**Describe the bug**
related to https://github.com/NVIDIA/spark-rapids/pull/3770

rapids_databricks_nightly-dev-github build ID 212

```
[2021-11-09T03:30:21.638Z] ../../src/main/python/hash_aggregate_test.py:1163: 
[2021-11-09T03:30:21.638Z] _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
[2021-11-09T03:30:21.638Z] 
[2021-11-09T03:30:21.638Z] df_fun = <function test_hash_groupby_approx_percentile_long_repeated_keys.<locals>.<lambda> at 0x7f11f7f65320>
[2021-11-09T03:30:21.638Z] percentiles = [0.05, 0.25, 0.5, 0.75, 0.95]
[2021-11-09T03:30:21.638Z] conf = {'spark.rapids.sql.expression.ApproximatePercentile': 'true', 'spark.sql.adaptive.enabled': 'false'}
[2021-11-09T03:30:21.638Z] 
[2021-11-09T03:30:21.638Z]     def compare_percentile_approx(df_fun, percentiles, conf):
[2021-11-09T03:30:21.638Z]     
[2021-11-09T03:30:21.638Z]         # create SQL statements for exact and approx percentiles
[2021-11-09T03:30:21.638Z]         p_exact_sql = create_percentile_sql("percentile", percentiles)
[2021-11-09T03:30:21.638Z]         p_approx_sql = create_percentile_sql("approx_percentile", percentiles)
[2021-11-09T03:30:21.638Z]     
[2021-11-09T03:30:21.638Z]         def run_exact(spark):
[2021-11-09T03:30:21.638Z]             df = df_fun(spark)
[2021-11-09T03:30:21.638Z]             df.createOrReplaceTempView("t")
[2021-11-09T03:30:21.638Z]             return spark.sql(p_exact_sql)
[2021-11-09T03:30:21.638Z]     
[2021-11-09T03:30:21.638Z]         def run_approx(spark):
[2021-11-09T03:30:21.638Z]             df = df_fun(spark)
[2021-11-09T03:30:21.638Z]             df.createOrReplaceTempView("t")
[2021-11-09T03:30:21.638Z]             return spark.sql(p_approx_sql)
[2021-11-09T03:30:21.638Z]     
[2021-11-09T03:30:21.638Z]         # run exact percentile on CPU
[2021-11-09T03:30:21.638Z]         exact = run_with_cpu(run_exact, 'COLLECT', _approx_percentile_conf)
[2021-11-09T03:30:21.638Z]     
[2021-11-09T03:30:21.638Z]         # run approx_percentile on CPU and GPU
[2021-11-09T03:30:21.638Z]         approx_cpu, approx_gpu = run_with_cpu_and_gpu(run_approx, 'COLLECT', _approx_percentile_conf)
[2021-11-09T03:30:21.638Z]     
[2021-11-09T03:30:21.638Z]         assert len(exact) == len(approx_cpu)
[2021-11-09T03:30:21.638Z]         assert len(exact) == len(approx_gpu)
[2021-11-09T03:30:21.638Z]     
[2021-11-09T03:30:21.638Z]         for i in range(len(exact)):
[2021-11-09T03:30:21.638Z]             cpu_exact_result = exact[i]
[2021-11-09T03:30:21.638Z]             cpu_approx_result = approx_cpu[i]
[2021-11-09T03:30:21.638Z]             gpu_approx_result = approx_gpu[i]
[2021-11-09T03:30:21.638Z]     
[2021-11-09T03:30:21.638Z]             # assert that keys match
[2021-11-09T03:30:21.638Z]             assert cpu_exact_result['k'] == cpu_approx_result['k']
[2021-11-09T03:30:21.639Z]             assert cpu_exact_result['k'] == gpu_approx_result['k']
[2021-11-09T03:30:21.639Z]     
[2021-11-09T03:30:21.639Z]             # extract the percentile result column
[2021-11-09T03:30:21.639Z]             exact_percentile = cpu_exact_result['the_percentile']
[2021-11-09T03:30:21.639Z]             cpu_approx_percentile = cpu_approx_result['the_percentile']
[2021-11-09T03:30:21.639Z]             gpu_approx_percentile = gpu_approx_result['the_percentile']
[2021-11-09T03:30:21.639Z]     
[2021-11-09T03:30:21.639Z]             if exact_percentile is None:
[2021-11-09T03:30:21.639Z]                 assert cpu_approx_percentile is None
[2021-11-09T03:30:21.639Z]                 assert gpu_approx_percentile is None
[2021-11-09T03:30:21.639Z]             else:
[2021-11-09T03:30:21.639Z]                 assert cpu_approx_percentile is not None
[2021-11-09T03:30:21.639Z] >               assert gpu_approx_percentile is not None
[2021-11-09T03:30:21.639Z] E               assert None is not None
[2021-11-09T03:30:21.639Z] 
[2021-11-09T03:30:21.639Z] ../../src/main/python/hash_aggregate_test.py:1263: AssertionError
```

**Comment by pxLi** [comment]
@andygrove can you help take a look? thanks~

**Comment by andygrove** [comment]
@pxLi How can I confirm that the cudf jar in this run included https://github.com/rapidsai/cudf/pull/9537 which should be the fix for this issue?

**Comment by andygrove** [comment]
I just tested with Databricks 7.3 and I cannot reproduce the issue.

```
[gw0] [ 50%] PASSED ../../src/main/python/hash_aggregate_test.py::test_hash_groupby_approx_percentile_long_repeated_keys[false] 
[gw1] [100%] PASSED ../../src/main/python/hash_aggregate_test.py::test_hash_groupby_approx_percentile_long_repeated_keys[true] 
```

**Comment by pxLi** [comment]
> @pxLi How can I confirm that the cudf jar in this run included [rapidsai/cudf#9537](https://github.com/rapidsai/cudf/pull/9537) which should be the fix for this issue?

the commit info of cudf jar can be simply fetched as
`unzip -p "cudf.jar" "cudf-java-version-info.properties"`
for nightly we always use the latest snapshot build from cudf-nightly pipeline (cudf_nightly-dev-github)

we also see failures in 
rapids_it-Dataproc build ID 350
rapids_it-AzureDatabricks-7.3  build ID 179

I checked the cudf jar in these failed builds, it is based on 3280be232d8379b73c7af0ef20503d7c88da9af2 which should include the #9537

**Comment by pxLi** [comment]
also seeing ../../src/main/python/hash_aggregate_test.py::test_hash_groupby_approx_percentile_long_repeated_keys[false] 

intermittently hanging and timeout some builds, e.g.
rapids_it-3.0.x-SNAPSHOT-dev-github build ID 256
rapids_it-3.2.x-SNAPSHOT-pre_release-github build ID 15

![image](https://user-images.githubusercontent.com/8086184/142334600-a86de13e-203d-4960-8cdd-9323b177a53f.png)
![image](https://user-images.githubusercontent.com/8086184/142334683-ca01aed1-7f6d-4f57-a977-b416a1e5487d.png)

**Comment by andygrove** [comment]
The issue appears to be that the last bucket in the t-digest data is corrupted (intermittently). Here are the values for the last 10 buckets in the t-digest when the test is passing:

## mean

```
1950 9882.88
1951 9884.85
1952 9888.76
1953 9890.01
1954 9922.68
1955 9967.43
1956 9975.05
1957 9987.18
1958 9994.150000000001
1959 0.0
```

## weight

```
1950 1.0
1951 1.0
1952 1.0
1953 1.0
1954 1.0
1955 1.0
1956 1.0
1957 1.0
1958 2.0
1959 0.0
```

Here are the last 10 values when the test is failing. The last entry for both `mean` and `weight` is incorrect.

## mean

```
1950 9882.88
1951 9884.85
1952 9888.76
1953 9890.01
1954 9922.68
1955 9967.43
1956 9975.05
1957 9987.18
1958 9994.150000000001
1959 2.2310560923097894E208
```

## weight

```
1950 1.0
1951 1.0
1952 1.0
1953 1.0
1954 1.0
1955 1.0
1956 1.0
1957 1.0
1958 2.0
1959 5.932533616687933E276
```

**Comment by mythrocks** [comment]
If I'm not mistaken, the last few entries in the `mean` list are worrying, for the "passing" case:
```
1957 9987.18
1958 9994.150000000001
1959 0.0
```
Aren't these sorted in increasing value of mean? `0.0` would be wrong then, no?

**Comment by sameerz** [comment]
`approx_percentile` is off by default, we are going to document this issue (PR #4173) and fix in 22.02.
